mlpipeline/pipeline/graph_utils.py

```diff
@@ -0,0 +1,240 @@
+        outputs: set
+            The set of pipeline outputs
+
+        Returns
+        -------
+        dict
+            graph representation of components
+
+        Raises
+        ------
+        RuntimeError
+            If the inputs of one component don"t match with the output of the component it depends on
+            OR inputs and outputs in incorrect format
+        """
+
+        graph = {}
+        for component_id in components:
+            dependencies = components[component_id].dependencies
+            if component_id not in graph:
+                graph[component_id] = set()
+
+            for dependency in dependencies:
+                if dependency not in graph:
+                    graph[dependency] = set()
+
+                if self._verify_edge(components, inputs, outputs, dependency, component_id):
+                    graph[dependency].add(component_id)
+                else:
+                    L.error("Incorrect graph, shutting down")
+                    raise RuntimeError("Incorrect link from {} to {}, inputs don't match the outputs".format(dependency, component_id))
+
+        if len(outputs) > 0:
+            graph["outputs"] = set()
+
+        for output in outputs:
+            output_split = output.split(".")
+            if len(output_split) == 1:
+                if output_split not in inputs:
+                    L.error("Incorrect outputs, shutting down")
+                    raise RuntimeError("Incorrect outputs {}".format(outputs))
+
+                graph["inputs"].add("outputs")
+
+            elif len(output_split) == 2:
+                if self._verify_edge(components, inputs, outputs, output_split[0], "outputs"):
+                    graph[output_split[0]].add("outputs")
+                else:
+                    L.error("Incorrect outputs, shutting down")
+                    raise RuntimeError("Incorrect outputs {}".format(outputs))
+
+            else:
+                L.error("Incorrect inputs, shutting down")
+                raise RuntimeError("Incorrect input {}, should be <component_id>.<input_name>".format(output))
+
+        return graph
+
+    def _get_topological_order(self, graph: dict) -> list:
+        """Gets the dependency graph topologically sorted and creates an order of components to be executed in optimal order.
+        Checks if the graph contains a cycle (if output of one component is referenced somewhere in its" deps)
+
+        Parameters
+        ----------
+        graph : dict
+            Components dependency graph
+
+        Returns
+        -------
+        list
+            running order of the graph
+
+        Raises
+        ------
+        RuntimeError
+            If the graph contains a cycle
+        """
+
+        visited = set()
+        running_order = []
+        for component_id in graph:
+            if component_id not in visited:
+                self._get_topological_order_util(component_id, graph, visited, running_order)
```

Comment:
I think if you marked the nodes visited by this particular DFS call (e.g. by adding to another set) then you could detect cycles earlier (and probably easier) in  `_get_topological_order_util` just by checking that each subsequent node you visit is not marked as visited.